### PR TITLE
server: make region statistic loss some precision

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

### What problem does this PR solve?

In the current situation, a change in traffic will trigger the update of `save cache` in the region heartbeat, which will cause the heartbeat to become more stressful. In fact, this part of the statistical data can work normally as long as it is within a certain accuracy.

### Check List

- Unit test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has configuration change
- Has HTTP API interfaces change (Don't forget to [add the declarative for API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tidb-ansible`](https://github.com/pingcap/tidb-ansible):
- Need to cherry-pick to the release branch

### Release note

```release-note
add `flow-loss-precision` to reduce the statistics update overhead
```

### CODE REVIEW

File: server/core/region_option.go [hunk 1]
```diff
@@ -35,6 +35,19 @@ func WithDownPeers(downPeers []*pdpb.PeerStats) RegionCreateOption {
...
 
+// WithFlowLossPrecision applys the lossPrecision for the region flows.
+func WithFlowLossPrecision(lossPrecision uint64) RegionCreateOption {
+	return func(region *RegionInfo) {
+		if lossPrecision == 0 {
+			return
+		}
+		region.writtenBytes = (region.writtenBytes / lossPrecision) * lossPrecision
+		region.readBytes = (region.readBytes / lossPrecision) * lossPrecision
+		region.writtenKeys = (region.writtenKeys / lossPrecision) * lossPrecision
```
[the remark this answers is not in the document]
Comment (nolouch) [reply server/core/region_option.go]: I have ignored the keys, PTAL.

File: server/config/config.go [hunk 2]
```diff
@@ -1070,7 +1071,10 @@ type PDServerConfig struct {
 	// There are some values supported: "auto", "none", or a specific address, default: "auto"
 	DashboardAddress string `toml:"dashboard-address" json:"dashboard-address"`
 	// TraceRegionFlow the option to update flow information of regions
+	// TODO: deprecate
 	TraceRegionFlow bool `toml:"trace-region-flow" json:"trace-region-flow,string"`
+	//FlowRoundStep is the step used to round to nearest.
```
Comment (rleungx) [inline server/config/config.go:1076]:
```suggestion
	// FlowRoundStep is the step used to round to the nearest.
```

File: server/core/region.go [hunk 3]
```diff
@@ -396,11 +397,27 @@ func (r *RegionInfo) GetBytesRead() uint64 {
 	return r.readBytes
 }
 
+// GetRoundBytesRead returns the read bytes of the region.
+func (r *RegionInfo) GetRoundBytesRead() uint64 {
+	if r.flowRoundStep == 0 {
+		return r.readBytes
+	}
+	return (r.readBytes / r.flowRoundStep) * r.flowRoundStep
```
Comment (rleungx) [inline server/core/region.go:405]: How about using shift operation?

File: server/config/config.go [hunk 4]
```diff
@@ -1090,6 +1094,9 @@ func (c *PDServerConfig) adjust(meta *configMetaData) error {
 	if !meta.IsDefined("trace-region-flow") {
 		c.TraceRegionFlow = defaultTraceRegionFlow
 	}
+	if !meta.IsDefined("flow-loss-precision") {
```
Comment (rleungx) [inline server/config/config.go:1097]: How about keeping consistent with the variable?

File: server/config/config.go [hunk 5]
```diff
@@ -1070,7 +1071,10 @@ type PDServerConfig struct {
 	// There are some values supported: "auto", "none", or a specific address, default: "auto"
 	DashboardAddress string `toml:"dashboard-address" json:"dashboard-address"`
 	// TraceRegionFlow the option to update flow information of regions
+	// TODO: deprecate
```
Comment (rleungx) [inline server/config/config.go:1074]: Can we directly mark it as deprecated?

Comment (nolouch) [reply server/config/config.go:1074]: some users already relies on it, I need to do some compatibility processing. will do it in another PR.

**Comment by nolouch** [comment]
PTAL @rleungx @HunDunDM

File: server/config/config.go (hunk 5 above)
Comment (nolouch) [reply server/config/config.go:1074]: test

**Review by Yisaer — APPROVED** [review]

File: server/config/config.go [hunk 6]
```diff
@@ -225,11 +225,11 @@ const (
...
 	defaultLeaderPriorityCheckInterval = time.Minute
 
-	defaultUseRegionStorage = true
-	defaultTraceRegionFlow  = true
-	defaultFlowBucketsWidth = 512
-	defaultMaxResetTSGap    = 24 * time.Hour
-	defaultKeyType          = "table"
+	defaultUseRegionStorage  = true
+	defaultTraceRegionFlow   = true
+	defaultFlowRroundByDigit = 3
```
Comment (rleungx) [inline server/config/config.go:230]:
```suggestion
	defaultFlowRoundByDigit = 3
```

File: server/config/config.go [hunk 7]
```diff
@@ -1073,8 +1079,8 @@ type PDServerConfig struct {
 	// TraceRegionFlow the option to update flow information of regions
 	// TODO: deprecate
 	TraceRegionFlow bool `toml:"trace-region-flow" json:"trace-region-flow,string"`
-	// FlowBucketsWidth is the width used to discretization processing flow information
-	FlowBucketsWidth uint64 `toml:"flow-buckets-width" json:"flow-buckets-width"`
+	// FlowRroundByDigit is the width used to discretization processing flow information
```
Comment (rleungx) [inline server/config/config.go:1082]: the comment needs change

Comment (nolouch) [reply server/config/config.go]: updated, ptal.

File: pkg/errs/errno.go [hunk 8]
```diff
@@ -271,6 +271,11 @@ var (
 	ErrSemverNewVersion = errors.Normalize("new version error", errors.RFCCodeText("PD:semver:ErrSemverNewVersion"))
 )
 
+// config
+var (
```
Comment (rleungx) [inline pkg/errs/errno.go:275]: Why the error is `PD:filepath:ErrFilePathAbs`?

Comment (nolouch) [reply pkg/errs/errno.go:275]: updated. PTAL

File: server/core/region_option.go [hunk 9]
```diff
@@ -35,6 +35,16 @@ func WithDownPeers(downPeers []*pdpb.PeerStats) RegionCreateOption {
 	}
 }
 
+// WithFlowRoundByDigit set the digit, which use to round to the nearest number
+func WithFlowRoundByDigit(digit int) RegionCreateOption {
```
Comment (HunDunDM) [inline server/core/region_option.go:39]: Why not use `TraceRegionFlow` and `FlowRoundByDigit` to calculate `divisor`?

File: server/core/region.go [hunk 10]
```diff
@@ -396,11 +398,29 @@ func (r *RegionInfo) GetBytesRead() uint64 {
...
+}
+
 // GetBytesWritten returns the written bytes of the region.
 func (r *RegionInfo) GetBytesWritten() uint64 {
 	return r.writtenBytes
 }
 
+// GetRoundBytesWritten returns the written bytes of the region.
+func (r *RegionInfo) GetRoundBytesWritten() uint64 {
+	divisor := uint64(math.Pow10(r.flowRoundByDigit))
```
Comment (HunDunDM) [inline server/core/region.go:417]: A lot of repeated calculations.

File: server/core/region.go [hunk 11]
```diff
@@ -396,11 +398,29 @@ func (r *RegionInfo) GetBytesRead() uint64 {
 	return r.readBytes
 }
 
+// GetRoundBytesRead returns the read bytes of the region.
+func (r *RegionInfo) GetRoundBytesRead() uint64 {
+	divisor := uint64(math.Pow10(r.flowRoundByDigit))
```
Comment (HunDunDM) [inline server/core/region.go:403]: A lot of repeated calculations.

Comment (nolouch) [reply server/core/region.go:403]: it's just a table lookup  `math.Pow10`

File: server/grpc_service.go [hunk 12]
```diff
@@ -502,6 +502,7 @@ func (s *heartbeatServer) Recv() (*pdpb.RegionHeartbeatRequest, error) {
 // RegionHeartbeat implements gRPC PDServer.
 func (s *Server) RegionHeartbeat(stream pdpb.PD_RegionHeartbeatServer) error {
 	server := &heartbeatServer{stream: stream}
+	FlowRoundByDigit := s.persistOptions.GetPDServerConfig().FlowRoundByDigit
```
Comment (HunDunDM) [inline server/grpc_service.go:505]: Get `TraceRegionFlow` here.

File: server/cluster/cluster.go [hunk 13]
```diff
@@ -655,11 +655,10 @@ func (c *RaftCluster) processRegionHeartbeat(region *core.RegionInfo) error {
...
 			saveCache = true
 		}
-
-		if c.traceRegionFlow && (region.GetBytesWritten() != origin.GetBytesWritten() ||
-			region.GetBytesRead() != origin.GetBytesRead() ||
-			region.GetKeysWritten() != origin.GetKeysWritten() ||
-			region.GetKeysRead() != origin.GetKeysRead()) {
+		// Once flow has changed, will update the cache.
+		// Because keys and bytes are strongly related, only bytes are judged.
+		if c.traceRegionFlow && (region.GetRoundBytesWritten() != origin.GetRoundBytesWritten() ||
```
Comment (HunDunDM) [inline server/cluster/cluster.go:660]: Don't get `TraceRegionFlow` here.

Comment (HunDunDM) [reply server/cluster/cluster.go:660]: Will be adjusted in another PR.

**Review by HunDunDM — APPROVED** [review]

**Pull request merged**